aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp

```diff
@@ -0,0 +1,575 @@
+
+    auto input_indexr = [=](int64_t c, int64_t d, int64_t h, int64_t w) {
+      return grad_input_data + c * input_depth * input_height * input_width +
+          d * input_height * input_width + h * input_width + w;
+    };
+
+    int64_t id0, id1, ih0, ih1, iw0, iw1;
+    scalar_t d0lambda, d1lambda, h0lambda, h1lambda, w0lambda, w1lambda;
+    for (int64_t c = begin; c < end; c++) {
+      for (int64_t od = 0; od < output_depth; od++) {
+        compute_source_index_and_lambda(
+            id0, id1, d0lambda, d1lambda, depth_scale, od, input_depth, output_depth, align_corners);
+        for (int64_t oh = 0; oh < output_height; oh++) {
+          compute_source_index_and_lambda(
+              ih0, ih1, h0lambda, h1lambda, height_scale, oh, input_height, output_height, align_corners);
+          for (int64_t ow = 0; ow < output_width; ow++) {
+            compute_source_index_and_lambda(
+                iw0, iw1, w0lambda, w1lambda, width_scale, ow, input_width, output_width, align_corners);
+            scalar_t grad_output_value = grad_output_data[c * output_slice_size +
+                od *  output_height * output_width + oh * output_width + ow];
+            *input_indexr(c, id0, ih0, iw0) += d0lambda * h0lambda * w0lambda * grad_output_value; /* i000 */
+            *input_indexr(c, id0, ih0, iw1) += d0lambda * h0lambda * w1lambda * grad_output_value; /* i001 */
+            *input_indexr(c, id0, ih1, iw0) += d0lambda * h1lambda * w0lambda * grad_output_value; /* i010 */
+            *input_indexr(c, id0, ih1, iw1) += d0lambda * h1lambda * w1lambda * grad_output_value; /* i011 */
+            *input_indexr(c, id1, ih0, iw0) += d1lambda * h0lambda * w0lambda * grad_output_value; /* i100 */
+            *input_indexr(c, id1, ih0, iw1) += d1lambda * h0lambda * w1lambda * grad_output_value; /* i101 */
+            *input_indexr(c, id1, ih1, iw0) += d1lambda * h1lambda * w0lambda * grad_output_value; /* i110 */
+            *input_indexr(c, id1, ih1, iw1) += d1lambda * h1lambda * w1lambda * grad_output_value; /* i111 */
+          }
+        }
+      }
+    }
+  };
+
+  if (ndim == 3) {
+    // upsample linear 1d
+    at::parallel_for(0, channels, at::internal::GRAIN_SIZE / output_slice_size / 2, loop1d);
+  } else if (ndim == 4) {
+    // upsample bilinear 2d
+    at::parallel_for(0, channels, at::internal::GRAIN_SIZE / output_slice_size / 4, loop2d);
+  } else {
+    // upsample trilinear 3d
+    TORCH_INTERNAL_ASSERT(ndim == 5);
+    at::parallel_for(0, channels, at::internal::GRAIN_SIZE / output_slice_size / 8, loop3d);
+  }
+
+  if (!grad_input_.is_contiguous()) {
+    grad_input_.copy_(grad_input);
+  }
+}
+
+using scale_t = std::vector<c10::optional<double>>;
+void upsample_linear1d_kernel_impl(
+    Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    c10::optional<double> scales_w) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "upsample_linear1d", [&] {
```

Comment:
Why are you dispatching half type but not bfloat16? I though half arithmetic is too slow on the cpu

Reply:
Indeed it doesn't make much sense to dispatch to **half**. It's a purely copy/paste from the old code [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UpSampleLinear1d.cpp#L196).
Anyway I have all **half** dispatch in upsample_xxx, let me know if you have a second thought.